Collapsible sidebar

gabalafou opened the pull request:
Screenshot showing alignment guidelines:

![](https://github.com/user-attachments/assets/af21c597-ead4-49c3-b9c3-63dc59c26272)

gabalafou commented:
I made the following accessibility checks:

- [x] No animation if user has set `prefers-reduced-motion: reduce`
- [x] Screen reader can read both expanded and collapsed state of button
- [x] Focus rings and keyboard functionality work
- [x] Minimum 24 x 24 px hit area for mouse click

gabalafou commented:
Update: I have incorporated feedback from @smeragoel. The only thing left to do with this PR is to add a bare minimum of tests, and then I will mark it ready for review. I will not be able to get to writing tests until next week.

dprada commented:
I am really looking forward to this implementation. Thanks @gabalafou and @smeragoel. 👏🏻

drammock commented:
> Update: I have incorporated feedback from @smeragoel.

Note to whoever merges this: please add @smeragoel as `co-authored-by`

gabalafou commented:
Okay, I added tests. This is now ready for review.

trallard reviewed (dismissed):
Thanks for working on this @gabalafou and thanks for adding the tests.

I did not spot anything major, but left a message about reusing existing test sites to try and avoid adding too many sites, especially since we already have one for sidebars testing.

gabalafou commented on docs/user_guide/layout.rst:
```diff
@@ -354,15 +354,15 @@ By default, it has the following configuration:
 .. code-block:: python
 
     html_sidebars = {
-        "**": ["sidebar-nav-bs", "sidebar-ethical-ads"]
+        "**": ["sidebar-collapse", "sidebar-nav-bs"]
     }
 
```
My understanding is that this line in the docs is simply incorrect. My understanding is that the theme defaults are defined in [`theme.conf`](https://github.com/pydata/pydata-sphinx-theme/pull/2119/files/9c10dd11153187c172cfb995bd72a8308b9dea72..cc499126b58469847ed00c7dac58a124fc8b5235). 

If you take a look at that file in this PR, you should find:

```diff
-- sidebars = sidebar-nav-bs.html
++ sidebars = sidebar-collapse, sidebar-nav-bs
```

The template `sidebar-ethical-ads` is not and was not defined there.

However, in that same `theme.conf` file, you should also find:

```
primary_sidebar_end = sidebar-ethical-ads
```

It looks to me that at some point in the past, the ethical ads template was defined in the `html_sidebars` configuration. But then later a feature was added to PST so that the ethical ads would always be pinned to the bottom of the left sidebar. A new configuration variable was created for this "pin to bottom" area and added to theme.conf and it was set to include ethical ads by default. For reference, see [ENH: Add left sidebar end section to bottom #543](https://github.com/pydata/pydata-sphinx-theme/pull/543).

trallard replied:
I see. Yes, it seems the docs were not updated to reflect this change. Thanks for catching this.

gabalafou commented on docs/user_guide/layout.rst:
```diff
@@ -382,6 +382,8 @@ By default, it has the following templates:
       # ...
     }
 
+`sidebar-ethical-ads.html`` is a placement for ReadTheDocs's Ethical Ads (will only show up on ReadTheDocs).
```
No it's just to make the docs more accurate - see my response to your other question about ethical ads.

gabalafou commented on src/pydata_sphinx_theme/assets/styles/pydata-sphinx-theme.scss:
```diff
@@ -47,6 +47,7 @@
 @import "./components/navbar-links";
 @import "./components/page-toc";
 @import "./components/prev-next";
+@import "./components/sidebar-collapse";
```
Good catch!

gabalafou commented on src/pydata_sphinx_theme/theme/pydata_sphinx_theme/components/sidebar-collapse.html:
```diff
@@ -0,0 +1,13 @@
+    {#- Careful: the contents of this span element are read by JavaScript and passed to the Bootstrap tooltip.
+      If you modify the markup within this tag, be sure that the tooltip on the expand sidebar button still works
+      and doesn't have any extraneous whitespace. -#}
+    {{- _("Expand Sidebar") -}}
```
Ah, right, good catch. I had meant to do that. I think I should update the translations in this PR. Will push a commit that does so.

gabalafou commented on tests/sites/collapse_sidebar_button/Makefile:
I'll give that a try for now, but I want to kick off a discussion soon about our approach to using test sites.

drammock replied:
> I want to kick off a discussion soon about our approach to using test sites.

#2171 would be a good place to have that discussion.

gabalafou commented:
@trallard thank you for review! I think I responded and addressed all of your comments. The PR now needs another round of review and then hopefully it will be good to go! :)

gabalafou commented:
### About the failing check

The link check is expected to fail because there is page in the docs that contains [links to each of the component files in GitHub](https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/layout.html#built-in-components-to-insert-into-sections). This PR introduces a new component: `sidebar-collapse`. But the GitHub URL to this new component will not be live until this PR is merged.

But there was one unexpected thing. The link check [reports several gnu.org links](https://github.com/pydata/pydata-sphinx-theme/actions/runs/14045359136/job/39324854991?pr=2159) as broken. But I suspect those might just be flukes, because when I click the links, they work. The error message for all four of those links looks the same:

```
(HTTPConnectionPool(host='www.gnu.org', port=80): Max retries exceeded with url: /software/gettext/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fc13020ae70>: Failed to establish a new connection: [Errno 101] Network is unreachable')))
```

drammock commented:
> The link check [reports several gnu.org links](https://github.com/pydata/pydata-sphinx-theme/actions/runs/14045359136/job/39324854991?pr=2159) as broken.

those have been failing regularly for over a week now at least. Currently 🙈 but hopefully fixed by #2174

trallard commented on tests/sites/sidebars/section2/no-sidebar.rst:
Thanks for making the configuration explicit 🌻

trallard reviewed (approved):
Thanks @gabalafou 🚀  this is a rather exciting new feature.
Will wait for the CI to complete and we can merge this.

trallard commented:
> The link check [reports several gnu.org links](https://github.com/pydata/pydata-sphinx-theme/actions/runs/14045359136/job/39324854991?pr=2159) as broken.

> those have been failing regularly for over a week now at least. Currently 🙈 but hopefully fixed by https://github.com/pydata/pydata-sphinx-theme/pull/2174

These fail _only_ on CI (but not always), and they might be related to some `robots.txt` or something else, as' link check` uses `requests` under the hood.
I might add the `gnu` URLS to the list of skipped URL checks to avoid the flaky errors.

That aside, this PR looks ready to merge so will wait for the rest of the CI to complete and merge it 🚀

pllim commented:
Ooo thanks! What release will this be in?

gabalafou commented on src/pydata_sphinx_theme/assets/styles/sections/_sidebar-primary.scss:
```diff
@@ -16,10 +16,10 @@ $sidebar-padding-right: 1rem;
   border-right: 1px solid var(--pst-color-border);
   background-color: var(--pst-color-background);
-  overflow-y: auto;
+  overflow: hidden auto;
```
Note to self: this is the one change that I worry about having unintended consequences. I don't see any reason why the primary sidebar should show in the horizontal direction and show a horizontal scroll bar, but I may not be aware of some sites or configurations that need the primary sidebar to scroll along the x axis.

gabalafou replied:
(The reason why I don't think it needs horizontal scroll is that the section table of contents is configured to reflow and ethical ads are also responsive.)

trallard replied:
True, this is one such instance where we'd have horizontal overflow.
@12rambau would it be possible to see how the new feature would render for your API docs? I am curious on whether we will be hitting some unexpected behaviours.

12rambau replied:
I'm currently travelling with only my phone so I cannot really make checks outside of the github realm and building this documentation requires a new release from my side. I'll let you know once I'm back home and I can locally do a test

gabalafou replied:
> By design sphinx will use the full path to the method/class you want to describe which can become very very long if unchecked. 

Yeah, we can see that on the [PST API docs page](https://pydata-sphinx-theme.readthedocs.io/en/latest/api/pydata_sphinx_theme/edit_this_page/index.html) with the long method name `pydata_sphinx_theme.edit_this_page`. 

I believe the alternative would be to force the method name on one line which, if the primary sidebar is narrow enough, will cause part of the string to be hidden on the right side (in the sidebar's overflow). That means the user will have to scroll horizontally in order to see the rest of the name. That feels to me like it makes the table of contents even harder to absorb visually than if the method name is broken across two or more lines.

Here are two screenshots to show the difference.

### No wrapping

With the page a bit narrow and text wrapping turned off, the user can only make out "edit this pa":

<img width="1134" alt="" src="https://github.com/user-attachments/assets/2fb970c4-fe92-43fd-aeef-9b94a3bc9f5b" />

### With wrapping

With the page a bit narrow and with text wrapping enabled, the user can see the entire method name:

<img width="1134" alt="" src="https://github.com/user-attachments/assets/3898fcd8-c5f8-429a-900a-f587036eedf0" />

12rambau replied:
I like the text wrap even if it looks a bit chunky at least you see everything.

trallard replied:
I think the wrapping is fine, I do not expect this to be a very frequent situation, and if it causes problems we can reevaluate/iterate.

gabalafou commented:
I came back to this PR because I suddenly thought I had forgotten to include @smeragoel on the git commit co-authors, but then I realized that I didn't do the actual merge for this PR (@trallard did). Then I looked at the [commit](https://github.com/pydata/pydata-sphinx-theme/commit/21a1870c903198ea05bcdb303da365422c42ceba) and saw the following at the bottom and it made me happy:


```
Co-authored-by: Daniel McCloy <dan@mccloy.info>
Co-authored-by: Tania Allard <taniar.allard@gmail.com>
Co-authored-by: Smera Goel <sgoel@quansight.com>
```

Thanks @drammock, @trallard, @smeragoel for all of your help, you are a great team!

🌻
